### PR TITLE
#836 make va_profile_opt_in_out lambda read certificates from lambda layers and ACM

### LINKED ISSUE #836: Include the VA root certificates with the va_profile_opt_in_out lambda deployment
[va_profile_opt_in_out_lambda](https://github.com/department-of-veterans-affairs/notification-api/tree/master/lambda_functions/va_profile) needs the VA's root certificate chain to authenticate VA Profile's server when making PUT requests to their endpoint.  As currently implemented, the lambda reads the certificates from SSM Parameter Store.

A better approach might be to emulate the [user flows lambda packaging process](https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/workflows/lambda-functions.yml#L54) to make these root certificates part of the lambda deployment so they do not need to be read from outside the environment.

This ticket will need a corresponding infrastructure change to remove the SSM parameters.

### PR DESCRIPTION
**Pull request by kalbfled**
# Description

These changes support reading JWT related certificates from disk (via a lambda layer), obtaining the ALB certificate chain from ACM, and obtaining the ALB certificate's private key from SSM Parameter Store.  These changes get PUT requests, with mTLS, working for VA Profile integration.

Fixes #836 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] All unit tests pass
- [x] Manually make a POST request to the lambda, deployed in dev, using Profile's Swagger endpoint.  Inspect the CloudWatch logs and associated database table.
See [the CloudWatch log](https://console.amazonaws-us-gov.com/cloudwatch/home?region=us-gov-west-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Fproject-dev-va-profile-opt-in-out-lambda/log-events/2022$252F10$252F05$252F$255B$2524LATEST$255D342e0228f0b2410da5786072c2ede75e).

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 1]
```diff
@@ -25,120 +32,115 @@
 import psycopg2
 import ssl
 import sys
-from botocore.exceptions import ClientError
+from botocore.exceptions import ClientError, ValidationError
 from cryptography.x509 import Certificate, load_pem_x509_certificate
 from http.client import HTTPSConnection
-from typing import Optional, Tuple
+from random import choices
+from string import ascii_letters
```
Comment (kalbfled) [inline lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:39]: I need to remove these.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 2]
```diff
@@ -25,120 +32,115 @@
...
-        elif name == VA_CERT_NAME:
-            if va_cert is not None:
-                logger.warning("Parameter Store returned multiple values for %s.", name)
-            va_cert = parameter.get("Value")
-        elif not name:
-            logger.debug("Parameter Store returned a parameter without a name.")
-        else:
-            logger.debug("Parameter Store returned an unsolicited parameter: %s", name)
+should_make_put_request = NOTIFY_ENVIRONMENT == "test" or (VA_PROFILE_DOMAIN is not None and ssl_context is not None)
+if not should_make_put_request and NOTIFY_ENVIRONMENT != "test":
```
Comment (kalbfled) [inline lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:126]: I'm changing this.

**Review by EvanParish — APPROVED** [review]

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 3]
```diff
@@ -25,120 +32,113 @@
...
 
 
-def get_certificates_from_ssm() -> Tuple[Optional[Certificate], Optional[Certificate]]:
-    """
-    Query AWS SSM Parameter Store to get the VA Profile public certificate and the VA root
-    certificate chain.
-    """
-
-    assert not requested_certificates, "Don't call this function more than once."
+# TODO - make SSM call; delete this block
```
Comment (cris-oddball) [inline lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:77]: TODO == ticket, please

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 4]
```diff
@@ -25,120 +32,113 @@
...
 
 logger = logging.getLogger("VAProfileOptInOut")
 logger.setLevel(logging.DEBUG)
 
+CERTIFICATE_ARN = os.getenv("CERTIFICATE_ARN")
 OPT_IN_OUT_QUERY = """SELECT va_profile_opt_in_out(%s, %s, %s, %s, %s);"""
 NOTIFY_ENVIRONMENT = os.getenv("NOTIFY_ENVIRONMENT")
-# TODO - Make this an SSM call.  Consolidate all parameter calls.
+PRIVATE_KEY_PATH = os.getenv("PRIVATE_KEY_PATH")
+# TODO - Make this an SSM call.
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:48]: https://github.com/department-of-veterans-affairs/vanotify-infra/issues/387

**Review by cris-oddball — APPROVED** [review]

**Review by jakehova — APPROVED** [review]

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 5]
```diff
@@ -185,22 +181,14 @@ def va_profile_opt_in_out_lambda_handler(event: dict, context, worker_id=None) -
...
-        requested_certificates = True
+    global va_profile_public_cert
 
     # Authenticate the request from VA Profile.
     headers = event.get("headers", {})
     if not jwt_is_valid(headers.get("Authorization", headers.get("authorization", '')), va_profile_public_cert):
         return { "statusCode": 401 }
 
-    post_body = event["body"]
+    post_body = event.get("body")
```
Comment (kalbfled) [inline lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:191]: If the event doesn't contain "body", post_body is assigned None, and the handler exits on line 200.

Comment (cris-oddball) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:191]: Good catch!

**Review by k-macmillan — APPROVED** [review]

**Review by cris-oddball — APPROVED** [review]

**Review by EvanParish — APPROVED** [review]

**Pull request merged**